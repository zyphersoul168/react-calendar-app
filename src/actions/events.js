import Swal from "sweetalert2";
import { types } from "../actionTypes/types";
import { fetchWithToken } from "../helpers/fetch";
import { parseEvents } from "../helpers/parseEvents";

// async actions
export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;

    try {
      const resp = await fetchWithToken("events", event, "POST");
      const body = await resp.json();

      if (body.ok) {
        event._id = body.event._id;
        event.user = {
          _id: uid,
          name: name,
        };
        console.log(event);
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken("events");
      const body = await resp.json();
      const events = parseEvents(body.events);
      dispatch(eventLoad(events));
      
    } catch (error) {
      console.log(error);
    }
  };
};

export const eventStartUpdate = (event) => {
  return async(dispatch) => {
    try {
      const resp =  await fetchWithToken(`events/${event._id}`, event, "PUT");
      const body = await resp.json();

      if(body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  }
}

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    const { _id } = getState().calendar.activeEvent;

    try {
      const resp =  await fetchWithToken(`events/${_id}`, {}, "DELETE");
      const body = await resp.json();

      if(body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  }
}

// sync actions
export const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActive = () => ({
  type: types.eventClearActive,
});

export const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventLoad = (events) => ({
  type: types.eventLoad,
  payload: events,
}) 

export const eventLogout = () => ({
  type: types.eventLogout,
});
