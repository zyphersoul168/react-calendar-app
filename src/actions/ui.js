import { types } from "../actionTypes/types";

export const uiOpenModal = () => ({
  type: types.uiOpenModal,
});

export const uiCloseModal = () => ({
  type: types.uiCloseModal,
});
