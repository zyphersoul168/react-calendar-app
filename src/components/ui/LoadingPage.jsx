import React from "react";

export const LoadingPage = () => {
  return (
    <div>
      <h1 className="text-3xl">Loading...</h1>
    </div>
  );
};
