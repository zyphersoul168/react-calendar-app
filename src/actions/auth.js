import Swal from "sweetalert2";
import { types } from "../actionTypes/types";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const response = await fetchWithoutToken(
      "auth/login",
      { email, password },
      "POST"
    );
    const body = await response.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startSignup = (name, email, password) => {
  return async (dispatch) => {
    const response = await fetchWithoutToken(
      "auth/signup",
      { name, email, password },
      "POST",
    );

    const body = await response.json();

    if(body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  }
};

export const startChecking = () => {
  return async (dispatch) => {
    const response = await fetchWithToken("auth/resignup");

    const body = await response.json();

    if(body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      dispatch(setCheckingFinish());
    }
  }
}

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(eventLogout());
    dispatch(logout());
  };
};

// sync actions
export const setCheckingFinish = () => ({
  type: types.authCheckingFinish,
});

export const login = (user) => ({
  type: types.authLogin,
  payload: user,
}); 

export const logout = () => ({
  type: types.authLogout,
});