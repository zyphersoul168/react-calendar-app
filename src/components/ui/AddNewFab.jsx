import React from "react";
import { PlusIcon } from "@heroicons/react/solid";
import { useDispatch } from "react-redux";
import { uiOpenModal } from "../../actions/ui";

export const AddNewFab = () => {
  const dispatch = useDispatch();

  const handleClick = () => {
    dispatch(uiOpenModal());
  };

  return (
    <button
      onClick={handleClick}
      className="bg-purple-300 hover:bg-purple-400 p-4 absolute bottom-3 right-3 rounded-full z-10 text-gray-800 transition ease-in"
    >
      <PlusIcon className="w-6 h-6" />
    </button>
  );
};
