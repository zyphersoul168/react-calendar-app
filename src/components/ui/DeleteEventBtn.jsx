import React from "react";
import { TrashIcon } from "@heroicons/react/solid";
import { useDispatch } from "react-redux";
import { eventStartDelete } from "../../actions/events";
import { uiCloseModal } from "../../actions/ui";

export const DeleteEventBtn = () => {
  const dispatch = useDispatch();
  
  const handleDelete = () => {
    dispatch(eventStartDelete());
    dispatch(uiCloseModal())
  };
  
  return (
    <button onClick={handleDelete} className="bg-red-400 hover:bg-red-500 p-2.5 font-bold rounded-md text-gray-50 flex justify-center w-full transition ease-in">
      <TrashIcon className="mt-0.5 mr-1 w-6 h-6" />
      <span>Delete</span>
    </button>
  );
};
