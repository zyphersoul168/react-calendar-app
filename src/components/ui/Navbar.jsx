import React from "react";
import { LogoutIcon } from "@heroicons/react/outline";
import { useSelector, useDispatch } from "react-redux";

import { startLogout } from "../../actions/auth";

export const Navbar = () => {
  // const btn = document.querySelector("button.mobile-menu-button");
  const menu = document.querySelector(".mobile-menu");

  const { name } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <nav className="bg-gray-900 shadow-lg">
      <div className="max-w-10xl mx-auto px-4 py-3 ">
        <div className="flex justify-between">
          <div className="flex space-x-7">
            {/* <!-- Primary Navbar items --> */}
            <div className="hidden md:flex items-center space-x-4">
              <p className="py-4 px-2 text-gray-50">Hello {name}! :)</p>
            </div>
          </div>
          {/* <!-- Secondary Navbar items --> */}
          <div className="hidden md:flex items-center space-x-3 ">
            <button
              onClick={handleLogout}
              className="inline-flex py-2 px-2 font-bold text-gray-50 bg-red-400 hover:bg-red-600 rounded transition duration-300"
            >
              <LogoutIcon className="mt-0.5 mr-1 h-5 w-5" /> Logout
            </button>
          </div>
          {/* <!-- Mobile menu button --> */}
          <div className="md:hidden flex items-center">
            <button
              className="outline-none mobile-menu-button"
              onClick={() => menu.classList.toggle("hidden")}
            >
              <svg
                className=" w-6 h-6 text-gray-200 hover:text-indigo-300"
                x-show="!showMenu"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                viewBox="0 0 24 24"
                stroke="currentColor"
              >
                <path d="M4 6h16M4 12h16M4 18h16"></path>
              </svg>
            </button>
          </div>
        </div>
      </div>
      {/* <!-- mobile menu --> */}
      <div className="hidden mobile-menu">
        <ul className="pb-5">
          <li className="active">
            <p className="block text-md px-2 py-4 text-gray-50 font-semibold">
              Hello {name}! :)
            </p>
          </li>
          <li>
            <button
              onClick={handleLogout}
              className="ml-2 inline-flex p-3 font-bold text-gray-50 bg-red-400 hover:bg-red-600 rounded transition duration-300"
            >
              <LogoutIcon className="mt-0.5 mr-1 h-5 w-5" /> Logout
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};
