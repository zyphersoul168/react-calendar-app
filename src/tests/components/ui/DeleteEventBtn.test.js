import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";

import { DeleteEventBtn } from "../../../components/ui/DeleteEventBtn";
import { eventStartDelete } from "../../../actions/events";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};
const store = mockStore(initialState);

// mocks
store.dispatch = jest.fn();
jest.mock("../../../actions/events", () => ({
  eventStartDelete: jest.fn(),
}));

const wrapper = mount(
  <Provider store={store}>
    <DeleteEventBtn />
  </Provider>
);

describe("Tests on delete event button", () => {
  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch eventStartDelete action on click", () => {
    wrapper.find("button").prop("onClick")();

    expect(eventStartDelete).toHaveBeenCalled();
  });
});
