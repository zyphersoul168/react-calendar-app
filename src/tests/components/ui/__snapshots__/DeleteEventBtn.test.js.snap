// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on delete event button should match snapshot 1`] = `
<button
  className="bg-red-400 p-2.5 border font-bold rounded-md text-alabaster flex justify-center w-full transition ease-in"
  onClick={[Function]}
>
  <svg
    className="mt-0.5 mr-1 w-6 h-6"
    fill="currentColor"
    viewBox="0 0 20 20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
      fillRule="evenodd"
    />
  </svg>
  <span>
    Delete
  </span>
</button>
`;
